Actually deprecate long standing features

nicoddemus opened the pull request:
This issues proper warnings for:

* Compatible properties
* `cached_setup`
* Using special named objects to customize node objects during class creation.

This fixes #3616. It also supports a new type of changelog entry, "deprecations", so we can show `Deprecations` separated from `Removals`.

@RonnyPfannschmidt are other features you remember of that we should issue deprecation warnings? Would love to add to them as well.

Also it follows that we should not merge #3898, as now we have a proper way to turn the warnings into errors in pytest-4.0, to effectively remove them in 4.1. 👍

RonnyPfannschmidt commented on src/_pytest/nodes.py:
```diff
@@ -61,11 +62,13 @@ def __get__(self, obj, owner):
-        #     "usage of {owner!r}.{name} is deprecated, please use pytest.{name} instead".format(
-        #         name=self.name, owner=type(owner).__name__),
-        #     PendingDeprecationWarning, stacklevel=2)
+        warnings.warn(
```
i had massive issue about one part of those warnings being triggered from pytest internals by accident which is why i deffered it on the old warnings system

nicoddemus replied:
`catch_warnings` to the rescue. 😁 

It won't show the warnings for this specific access, but I guess that's OK because the whole class will be kicked out anyway. 👍

RonnyPfannschmidt commented on src/_pytest/python.py:
```diff
@@ -800,7 +800,7 @@ def collect(self):
                 )
             seen[name] = True
-            values.append(self.Function(name, self, args=args, callobj=call))
+            values.append(Function(name, self, args=args, callobj=call))
```
this is a breaking change, not doing it is why i had massive trouble with the warnings

nicoddemus commented:
> please move the actual warning texts/instances to the deprecation modules

Argh I missed that point. Definitely, will do.

> also we should take a look at creating a minimal tool for creating warnings that will be formatted at display time

Can you please clarify what you mean here?

RonnyPfannschmidt commented:
we do have more and more warnings where we format in code - its not nice to use them in code as one has to import the message and still pick the correct category

instead of `warn(SomeWarningType(CONSTANT.format(foo)))` i want to write `warn(CONSTANT.format(foo))`

nicoddemus commented:
> we do have more and more warnings where we format in code - its not nice to use them in code as one has to import the message and still pick the correct category

I suppose we can have `format` method in our warnings that returns a new instance of that warning, but applying the formatting.

Could be simply:

```python
class PytestWarning:
    def format(self, *args, **kwargs):
        return type(self)(self.args[0].format(*args, **kwargs)

FUNCARG_PREFIX = RemovedInPytest4Warning(
    '{name}: declaring fixtures using "pytest_funcarg__" prefix is deprecated')
```

Usage:

```python
warn(FUNCARG_PREFIX.format(name=f.name))
```

RonnyPfannschmidt commented:
instead i would like to see a 
```
@attr.s
class UnformattedWarning(object):
    category = attr.ib()
    template = attr.ib()

    def format(self, **kw):
        return self.category(self.template.format(**kw))

FUNCARG_PREFIX = UnformattedWarning(
   RemovedInPytest4Warning,
   '{name}: declaring fixtures using "pytest_funcarg__" prefix is deprecated',
)
```

that way we have a much better indicator of missuse

nicoddemus commented:
Oh great idea. I will implement it then. 👍

nicoddemus commented:
Implemented @RonnyPfannschmidt! The result is much better, thanks for the suggestion.

I also found one warning which was using the wrong category, I fixed and noted in the CHANGELOG. 👍

RonnyPfannschmidt commented on src/_pytest/deprecated.py:
```diff
@@ -4,11 +4,32 @@
 
+
+@attr.s
+class UnformattedWarning(object):
```
we might want to move this to warning_types - then this module can be only constants

nicoddemus replied:
Done!
